### ISSUE TITLE
Fix filtering step

### ISSUE THREAD

**Issue by ezylstra** [Big-Biodiversity-Collaborative/SwallowtailClimateChange#66]
Some occurrence records have no climate data, or share a raster cell with a second occurrence record (eg, P. machaon).  We're notified of the problem when we run a maxent with ENMeval. Possible solution: do NA and shared cell filtering with both terra and raster objects.

**Comment by jcoliver** [comment]
TL;DR: I think this problem lies with `ENMeval::ENMevaluate()` and one that we are unlikely to fix. Thoughts, @ezylstra ? I have refrained from pushing the updated gbif-2-filter.R script, as I think it adds compute that doesn't solve our problem.

Working on this, testing with _P. machaon_. Implemented climate NA removal and thinning via terra and raster approaches, but did not fix the issue when running MaxEnt in butterfly-sdms-spatialCV.R. Two observations removed (one for NA climate, one for shared cell.)

| row | gbifID | Longitude | Latitude | Reason |
|:-|:-|:-|:-|:-|
|20 | 2466530611 | -112.0000 | 68.25000 | Climate data NA |
|368 | 3350920166 | -108.9200 | 49.50000 | Shared cell with another point |

Likely related to the fact that these points are on grid cell intersections:

```
bio1 <- terra::rast(x = "data/wc2-1/bio1.tif")

# The point excluded due to climate NA:
bio1_crop <- terra::crop(x = bio1, y = terra::ext(-112.2, -111.8, 68.15, 68.3))
plot(bio1_crop)
points(x = -112, y = 68.25, col = "red", pch = 19)

# The point excluded due to sharing a cell with another point:
bio1_crop <- terra::crop(x = bio1, y = terra::ext(-109, -108.8, 49.4, 49.6))
plot(bio1_crop)
# The point that gets removed (on cell boundary)
points(x = -108.92, y = 49.5, col = "red", pch = 19)
# The point that is retained
points(x = -108.9536, y = 49.50181, col = "black", pch = 19)
```

Point removed due to missing climate data:

![image](https://github.com/Big-Biodiversity-Collaborative/SwallowtailClimateChange/assets/1939813/9e265db7-77ff-459b-a8cb-029febafaca9)

Point removed (red) because there is another point (black) in the same cell:

![image](https://github.com/Big-Biodiversity-Collaborative/SwallowtailClimateChange/assets/1939813/da5e6851-8ed5-41d9-a2e7-a85c52d09e16)

Furthermore, when running the line of code in `ENMeval::ENMevaluate()` that produces the warning on the original data set (i.e. the `occs` object from lines 83-85 in butterfly-sdms-spatialCV.R), it **does not** indicate any duplicates (see lines 444-446 in https://github.com/jamiemkass/ENMeval/blob/199bf0181716b25ea5033be16ed8c6efadcfbd95/R/ENMevaluate.R#L447).

Just looking at the three original points, they appear to (1) be in cells with climate data and (2) be in different cells when using the `raster::extract()` function (below) and `terra::extract()` function (results not shown).

```
bio1_raster <- raster::raster(x = "data/wc2-1/bio1.tif")
unfiltered <- read.csv(file = "data/gbif/downloaded/papilio_machaon-gbif.csv")
cell_nums <- raster::extract(x = bio1_raster, 
                             y = unfiltered[, c("longitude", "latitude")], 
                             cellnumbers = TRUE)
cell_nums <- data.frame(cell_nums)
# Add GBIF ids so we can pull out specific observations
cell_nums$gbifID <- unfiltered$gbifID
cell_nums[cell_nums$gbifID %in% c(2466530611, 3350920166, 3335373107), ]
```
| ID | cells | bio1 | gbifID |
|:-|:-|:-|:-|
| 115 | 1775786  | 3.803565 | 3335373107 |
| 319  | 471817 | -4.358409 | 2466530611 |
| 846 | 1778690  | 4.083460 | 3350920166 |

So I'm just stumped. :-/

**Comment by ezylstra** [comment]
@jcoliver Thanks for looking into this in depth!  We do have a workaround if necessary, but I'll look more at things tomorrow morning just in case (though I'm sure you're right).  I'll let you know either way.

**Comment by ezylstra** [comment]
I think I solved it!  The problem isn't due to ENMeval, but actually occurs when cropping and masking the rasters.  If we add an argument: snap = "in", to the `terra::crop()` function, we don't run into problems with NA climate values or multiple occurrence records in one cell. This means that the `gbif-2-filter.R` script is fine as is. All changes will need to be made downstream (eg, in `template-CV.R`) when cropping rasters for a particular species. Given that, I'm going to go ahead and create new presence-absence datasets with the spatial folds.
